README.md: Document CMake build support

LDong-Arm opened the pull request:
Preceding PR: #26 

(The last commit belongs to this PR.)
Document how to build greentea-client in another CMake project and set up a stream of I/O.

LDong-Arm committed:
README.md: Document CMake build support

Document how to
* include greentea-client in another CMake project
* set up a stream of I/O
* build and run the provided examples

LDong-Arm commented on README.md:
```diff
@@ -68,3 +73,48 @@ You can find the greentea-client example application [here](./example/main.cpp).
 
 This generates an executable `./cmake_build/example/greentea-client-example`.
+
+# Adding greentea-client to a project
```
@Patater @rwalton-arm Please review whether this section is consistent with how we integrate greentea-client with mbed-os, thanks.

rwalton-arm replied:
One small issue that I can see: CMake isn't a "build system", it's a build system generator. The "build system" would be Ninja, make etc.

LDong-Arm replied:
Done. I changed it to simply "supports CMake".

Patater commented on README.md:
```diff
@@ -68,3 +73,48 @@ You can find the greentea-client example application [here](./example/main.cpp).
+
+greentea-client supports the CMake build system. To add greentea-client to a project, call
+[`add_subdirectory`](https://cmake.org/cmake/help/latest/command/add_subdirectory.html)
+pointing to the path to the client. Then link `greentea-client` to a CMake target that requires it with
```
```diff
- pointing to the path to the client.
+ pointing to the relative path of the `greentea-client` project directory.
```

LDong-Arm replied:
Done

Patater commented on README.md:
```diff
@@ -68,3 +73,48 @@ You can find the greentea-client example application [here](./example/main.cpp).
+greentea-client supports the CMake build system. To add greentea-client to a project, call
+[`add_subdirectory`](https://cmake.org/cmake/help/latest/command/add_subdirectory.html)
+pointing to the path to the client. Then link `greentea-client` to a CMake target that requires it with
+[`target_link_libraries`](https://cmake.org/cmake/help/latest/command/target_link_libraries.html).
```
Having examples of both `add_subdirectory(greentea-client)` and `target_link_libraries(my_app greentea-client)` would help with clarity.

LDong-Arm replied:
Added

Patater commented on README.md:
````diff
@@ -68,3 +73,48 @@ You can find the greentea-client example application [here](./example/main.cpp).
+    PRIVATE
+        custom_test_io.cpp
+)
+```
````
Could we make a custom IO example (our second example) and link to it instead of adding untested, may-go-stale code in the README?

LDong-Arm replied:
Added

LDong-Arm committed:
Add example to use custom IO

Now two examples are provided:
* example/stdio: Demonstrates the use of stdio for greentea-client
* example/custom_io: Demonstrates the use of a custom IO based on
regular files

Both are built by GitHub Actions for PR testing.

LDong-Arm commented:
@Patater @rwalton-arm Thanks for the suggestions, I've added an example to demonstrate a custom IO based on regular files, please review.

LDong-Arm commented on CMakeLists.txt:
```diff
@@ -32,10 +32,14 @@ if(GREENTEA_CLIENT_STDIO)
 
 if((CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME) AND BUILD_EXAMPLES)
-    add_subdirectory(example)
+    if(GREENTEA_CLIENT_STDIO)
+        add_subdirectory(examples/stdio)
+    else()
+        add_subdirectory(examples/custom_io)
+    endif()
```
CMake question: Is it possible to build two applications at once, with two different configurations (GREENTEA_CLIENT_STDIO ON/OFF) of a static libraries? I know it's doable if both applications use same configuration of the library...

Patater replied:
I'm not aware of a way to do that elegantly. Inelegantly, you could write a Makefile which calls `cmake` with each set of configuration options and builds the targets you want. Another approach is to specify two different `build` directories, one for each configuration set, and then e.g. `cmake --build build_stdio` and `cmake --build build_nostdio`.

LDong-Arm replied:
Fair enough. I guess building two versions of the same target in one go is not a supported/intended use case of CMake. I'll keep this unchanged.

Patater commented on examples/stdio/main.c:
```diff
@@ -14,12 +14,13 @@
 {
     const char *key = "demo_key";
-    const int value = 60;
+    const char *value = "60";
```
Why the change to a string? Shall we explain in a standalone commit's message?

LDong-Arm replied:
I converted this example to C, so overloads of `greentea_send_kv` are unavailable and the C version of it takes a string.

Maybe I should convert it back to C++, to allow demonstration of the full API?

LDong-Arm replied:
Done, I've converted the files back to C++

Patater replied:
I didn't pick up that C only allowed strings. I am okay with or without this change, just we need a commit message to explain it if we want to make the change.

LDong-Arm replied:
I've already converted it back to C++ because it allows us to use the full API. Maybe in the future we can send several KV pairs of different types (string, int, multiple ints for reporting passes & failures, ...) in our examples to demonstrate this. So does the current state of the PR look okay?

Patater replied:
Yup, all fine.
